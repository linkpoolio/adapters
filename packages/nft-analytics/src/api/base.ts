import { IFloorPrice } from '../models/floorprice'

export interface IFloorPrices {
  get: (props: any) => Promise<IFloorPrice>
}

export interface Base {
  floorprices: IFloorPrices
}
