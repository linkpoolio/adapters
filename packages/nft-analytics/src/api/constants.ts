export enum Provider {
  RARIFY = 'rarify',
}

export const providerToBaseUrl: ReadonlyMap<Provider, string> = new Map([
  [Provider.RARIFY, 'https://api.rarify.tech'],
])
