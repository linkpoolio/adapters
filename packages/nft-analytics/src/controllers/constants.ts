export enum Endpoint {
  FLOORPRICES = 'floorprices',
  TWAPS = 'twaps',
}

export enum RequestMethod {
  GET = 'get',
}
