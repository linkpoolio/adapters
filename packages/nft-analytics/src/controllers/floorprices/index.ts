import execute from './controller'
import { inputParameters } from './input'

const supportedEndpoints = ['floorprices']
const endpointResultPaths = {}
const description = 'This endpoint returns the floor price for a given NFT collection'

export { supportedEndpoints, endpointResultPaths, description, inputParameters, execute }
