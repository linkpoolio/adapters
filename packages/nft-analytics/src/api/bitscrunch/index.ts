import { Requester } from '@chainlink/ea-bootstrap'
import { Config, RequestConfig } from '@chainlink/types'
import { injector } from '@linkpool/shared'

import { Base } from '../base'
import { Provider as ProviderName, providerToBaseUrl } from '../constants'
import valuations from './valuations'

const Fetch = (config: Config) => {
  config.api.baseURL = providerToBaseUrl.get(ProviderName.BITSCRUNCH) as string
  config.api.headers['x-api-key'] = config.apiKey

  return async ({ url }: RequestConfig): Promise<any> => {
    const options = {
      ...config.api,
      url,
    }
    const response = await Requester.request<any>(options)

    return response
  }
}
const Provider = (config: Config): Base => {
  const fetch = Fetch(config)
  const endpoints = {
    valuations,
  }

  return injector(endpoints, fetch)
}

export default Provider
