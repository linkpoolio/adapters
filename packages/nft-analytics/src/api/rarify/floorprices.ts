import { util } from '@chainlink/ea-bootstrap'

import FloorPrice from '../../models/floorprice'
import type { IFloorPrice } from '../../models/floorprice'
import type { IFloorPrices } from '../base'
import { Provider } from '../constants'

export default (fetch): IFloorPrices => ({
  get: async ({ network, collectionAddress }): Promise<IFloorPrice> => {
    const url = util.buildUrlPath(`data/contracts/:network::collectionAddress/smart-floor-price`, {
      network,
      collectionAddress,
    })
    const response = await fetch({ url })
    return FloorPrice.Single(response.data, Provider.RARIFY)
  },
})
