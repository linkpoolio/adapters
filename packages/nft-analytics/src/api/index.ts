import { AdapterError, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

import rarify from './rarify'
import { Provider } from './constants'

export default (config: Config) => {
  const provider = util.getEnv('API_PROVIDER')

  switch (provider) {
    case Provider.RARIFY:
      return rarify(config)
    default: {
      const message = `Unsupported provider: ${provider}. Check API_PROVIDER env var`
      throw new AdapterError({
        statusCode: 200,
        message,
        cause: message,
      })
    }
  }
}
