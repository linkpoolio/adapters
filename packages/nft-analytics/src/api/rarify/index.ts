import { Requester } from '@chainlink/ea-bootstrap'
import { Config, RequestConfig } from '@chainlink/types'
import { injector } from '@linkpool/shared'

import { Base } from '../base'
import floorprices from './floorprices'
import { Provider as ProviderName, providerToBaseUrl } from '../constants'

const Fetch = (config: Config) => {
  config.api.headers['Authorization'] = `Bearer ${config.apiKey}`

  return async ({ url }: RequestConfig): Promise<any> => {
    const options = {
      ...config.api,
      baseURL: providerToBaseUrl.get(ProviderName.RARIFY) as string,
      url,
    }
    const response = await Requester.request<any>(options)

    return response
  }
}
const Provider = (config: Config): Base => {
  const fetch = Fetch(config)
  const endpoints = {
    floorprices,
  }

  return injector(endpoints, fetch)
}

export default Provider
