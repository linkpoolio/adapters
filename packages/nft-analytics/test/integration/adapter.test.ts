import http from 'http'
import type { AddressInfo } from 'net'
import process from 'process'
import request from 'supertest'
import type { SuperTest, Test } from 'supertest'

import { server as startServer } from '../../src'
import { Provider } from '../../src/api/constants'
import { testFloorprices } from './rarify/floorprices'

let oldEnv: NodeJS.ProcessEnv

export interface SuiteContext {
  server: http.Server
  req: SuperTest<Test>
}

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.CACHE_ENABLED = 'false'
  process.env.WARMUP_ENABLED = 'false'
  process.env.API_PROVIDER = Provider.RARIFY
  process.env.API_KEY = 'fake-api-key'
})

afterAll(() => {
  process.env = oldEnv
})

describe('execute', () => {
  const context: SuiteContext = {
    server: null,
    req: null,
  }

  beforeEach(async () => {
    context.server = await startServer()
    context.req = request(`localhost:${(context.server.address() as AddressInfo).port}`)
  })

  afterEach((done) => {
    context.server.close(done)
  })

  describe('rarify', () => {
    describe('floorprices endpoint', () => testFloorprices(context))
  })
})
