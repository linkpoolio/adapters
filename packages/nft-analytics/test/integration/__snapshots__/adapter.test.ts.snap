// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute rarify floorprices endpoint method: get returns a single floorprice 1`] = `
Object {
  "data": Object {
    "collection": Object {
      "contractAddress": "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
      "imageUrl": "https://rarify-public.s3.amazonaws.com/eth_logo.svg",
      "name": null,
      "symbol": null,
    },
    "floorPrice": Object {
      "amount": "77157500000000000000",
      "currency": "ETH",
      "units": "wei",
    },
    "network": "ethereum",
    "timestamp": 1659003693,
  },
  "jobRunID": 1,
  "statusCode": 200,
}
`;

exports[`execute rarify floorprices endpoint method: get returns an error response when the API errors too 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"floorprices\\",\\"method\\":\\"get\\",\\"collectionAddress\\":\\"0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d\\",\\"network\\":\\"ethereum\\"}}",
    "message": "Unexpected error requesting the 'floorprices' endpoint. Reason: AdapterError: Request failed with status code 500",
    "name": "AdapterError",
    "url": "https:/api.rarify.tech/data/contracts/ethereum:0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/smart-floor-price",
  },
  "jobRunID": 1,
  "providerStatusCode": 500,
  "status": "errored",
  "statusCode": 200,
}
`;
